lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -140,94 +145,229 @@
             else:
-                module.fail_json(msg='Could not detect supported package manager')
-    except Exception as e:
-        from traceback import format_tb
-        module.fail_json(msg='Failed to retrieve packages: %s' % to_text(e), exception=format_tb(sys.exc_info()[2]))
+                pkg['port_epoch'] = 0
+
+            if '_' in pkg['version']:
+                pkg['version'], pkg['revision'] = pkg['version'].split('_', 1)
+            else:
+                pkg['revision'] = 0
+
+        if 'vital' in pkg:
+            pkg['vital'] = bool(pkg['vital'])
+
+        return pkg
+
+
+class PORTAGE(CLIMgr):
+
+    CLI = 'qlist'
+    atoms = ['category', 'name', 'version', 'ebuild_revision', 'slots', 'prefixes', 'sufixes']
+
+    def list_installed(self):
+        rc, out, err = self.m.run_command(' '.join([self.cli, '-Iv', '|', 'xargs', '-n', '1024', 'qatom']), use_unsafe_shell=True)
+        if rc != 0 or err:
+            raise RuntimeError("Unable to list packages rc=%s : %s" % (rc, to_native(err)))
+        return out.splitlines()
+
+    def get_package_details(self, package):
+        return dict(zip(self.atoms, package.split()))
+
+
+class RPM(LibMgr):
+
+    LIB = 'rpm'
+
+    def list_installed(self):
+        return self.lib.TransactionSet().dbMatch()
+
+    def get_package_details(self, package):
+        return dict(name=package[self.lib.RPMTAG_NAME],
+                    version=package[self.lib.RPMTAG_VERSION],
+                    release=package[self.lib.RPMTAG_RELEASE],
+                    epoch=package[self.lib.RPMTAG_EPOCH],
+                    arch=package[self.lib.RPMTAG_ARCH],)
+
+
+def main():
+
+    # get supported pkg managers
+    PKG_MANAGERS = dict([(obj.__name__.lower(), obj) for obj in get_all_subclasses(PkgMgr) if obj not in(CLIMgr, LibMgr)])
+
+    # start work
+    module = AnsibleModule(argument_spec=dict(manager={'type': 'list', 'default': ['auto']}), supports_check_mode=True)
+    packages = {}
+    results = {'errors': []}
+    managers = module.params['manager']
+
+    if 'auto' in managers:
+        managers.extend(PKG_MANAGERS.keys())
+        managers.remove('auto')
+
+    unsupported = set(managers).difference(set(PKG_MANAGERS.keys()))
+    if unsupported:
+        module.fail_json(msg='Unsupported package managers requested: %s' % (', '.join(unsupported)))
+
+    found = 0
+    seen = set()
+    for pkgmgr in managers:
+
+        # avoid x2 procssing, but allow to keep order from user
+        if pkgmgr in seen:
+            continue
+        seen.add(pkgmgr)
+        try:
+            manager = PKG_MANAGERS[pkgmgr](module)
+            if manager.test():
+                found += 1
+                packages.update(manager.get_packages())
+        except Exception as e:
+            from traceback import format_tb
+            results['errors'].append({'msg': 'Failed to retrieve packages: %s' % to_text(e), 'exception': format_tb(sys.exc_info()[2])})
```

Comment:
There's some cornercases here which make me think we want to determine whether this is a warning or an error later.  A user can install package management software on any box but it might not work on that box.  That might not make a difference to the user at all.

Reply:
here i expect to find all kinds of corner cases, which is why i ignore this mostly by default, using warnings in stead might be fine